test/common/http/http1/codec_impl_test.cc

```diff
@@ -3904,5 +3880,301 @@ TEST_P(Http1ClientConnectionImplTest, InvalidResponseFirstCharacter) {
+TEST_P(Http1ServerConnectionImplTest, EOFDuringContentLengthBody) {
+  initialize();
+  InSequence s;
+
+  StrictMock<MockRequestDecoder> decoder;
+  Http::ResponseEncoder* response_encoder = nullptr;
+  EXPECT_CALL(callbacks_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder& encoder, bool) -> RequestDecoder& {
+        response_encoder = &encoder;
+        return decoder;
+      }));
+
+  EXPECT_CALL(decoder, decodeHeaders_(_, false));
+  EXPECT_CALL(decoder, decodeData(BufferStringEqual("foo"), false));
+  Buffer::OwnedImpl buffer("POST / HTTP/1.1\r\n"
+                           "content-length: 9\r\n\r\n"
+                           "foo");
+  auto status = codec_->dispatch(buffer);
+  EXPECT_EQ(0, buffer.length());
+  ASSERT_TRUE(status.ok());
+
+  EXPECT_CALL(decoder,
+              sendLocalReply(Http::Code::BadRequest, "Bad Request", _, _, "http1.codec_error"));
+  Buffer::OwnedImpl empty;
+  status = codec_->dispatch(empty);
+  EXPECT_TRUE(isCodecProtocolError(status));
+  EXPECT_EQ(status.message(), "http/1.1 protocol error: HPE_INVALID_EOF_STATE");
+  EXPECT_EQ("http1.codec_error", response_encoder->getStream().responseDetails());
+}
+
+// A read of zero bytes before Content-Length bytes of response body are read is an error.
+TEST_P(Http1ClientConnectionImplTest, EOFDuringContentLengthBody) {
+  initialize();
+  InSequence s;
+
+  StrictMock<MockResponseDecoder> decoder;
+  Http::RequestEncoder& request_encoder = codec_->newStream(decoder);
+  TestRequestHeaderMapImpl headers{{":method", "GET"}, {":path", "/"}, {":authority", "host"}};
+  EXPECT_TRUE(request_encoder.encodeHeaders(headers, true).ok());
+
+  EXPECT_CALL(decoder, decodeHeaders_(_, false));
+  EXPECT_CALL(decoder, decodeData(BufferStringEqual("foo"), false));
+  Buffer::OwnedImpl buffer("HTTP/1.1 200 OK\r\n"
+                           "content-length: 9\r\n\r\n"
+                           "foo");
+  auto status = codec_->dispatch(buffer);
+  EXPECT_EQ(0, buffer.length());
+  ASSERT_TRUE(status.ok());
+
+  Buffer::OwnedImpl empty;
+  status = codec_->dispatch(empty);
+  EXPECT_TRUE(isCodecProtocolError(status));
+  EXPECT_EQ(status.message(), "http/1.1 protocol error: HPE_INVALID_EOF_STATE");
+}
+
+// A request method requiring a body but without a Content-Length (or Transfer-Encoding: chunked)
+// header is an error.
+TEST_P(Http1ServerConnectionImplTest, NoContentLengthRequest) {
+  initialize();
+  InSequence s;
+
+  StrictMock<MockRequestDecoder> decoder;
+  Http::ResponseEncoder* response_encoder = nullptr;
+  EXPECT_CALL(callbacks_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder& encoder, bool) -> RequestDecoder& {
+        response_encoder = &encoder;
+        return decoder;
+      }));
+
+  if (parser_impl_ == Http1ParserImpl::BalsaParser) {
+    EXPECT_CALL(decoder,
+                sendLocalReply(Http::Code::BadRequest, "Bad Request", _, _, "http1.codec_error"));
+  } else {
+    EXPECT_CALL(decoder, decodeHeaders_(_, true));
+  }
```

Comment:
here and below: Whoa, nice! All of these differences make great sense (for Balsa).